coremltools/models/utils.py

```diff
@@ -2252,3 +2252,138 @@ def _make_second_chunk_prog(prog: _mil.Program, op_idx: int) -> _mil.Program:
+        - only an mlProgram is supported (not pipelines, not neural networks).
+
+    from_type:
+        The output type that should be changed from.
+
+    to_type:
+        The output type that will be used instead of all the outputs of `from_type` type.
+
+    function_names:
+        Optional list of function names where the output needs to be changed. If not specified, only the "main"
+        function will be updated.
+
+    Examples
+    --------
+    .. sourcecode:: python
+
+        from coremltools.models.model import MLModel
+        from coremltools.utils import change_array_output_type
+        from coremltools.proto.FeatureTypes_pb2 import ArrayFeatureType
+
+        model = MLModel("my_model.mlpackage")
+        updated_model = change_output_tensor_type(
+            ml_model=model,
+            from_type=ArrayFeatureType.FLOAT32,
+            to_type=ArrayFeatureType.FLOAT16,
+        )
+        updated_model.save("my_updated_model.mlpackage")
+    """
+    # We do the lazy import to prevent circular import
+    from coremltools.converters.mil.converter import mil_convert as _mil_convert
+    from coremltools.converters.mil import Var
+    from coremltools.converters.mil.mil.block import Function
+    from coremltools.proto.Model_pb2 import FeatureDescription
+
+    def _get_dtype(feature_type: int) -> str:
+        if feature_type == _ct.proto.FeatureTypes_pb2.ArrayFeatureType.FLOAT16:
+            return "fp16"
+        if feature_type == _ct.proto.FeatureTypes_pb2.ArrayFeatureType.FLOAT32:
+            return "fp32"
+        raise ValueError(f"invalid feature type: {feature_type}, supported only FLOAT16, FLOAT32")
+
+    def _get_output_var(var_name: str) -> _Optional[_Tuple[Function, Var]]:
+        for name in function_names:
+            func = prog.functions[name]
+            if var := next(iter([v for v in func.outputs if v.name == var_name]), None):
+                return func, var
+        return None
+
+    def _cast_output(feature_desc: FeatureDescription, feature_var: Var) -> None:
+        with feature_var.op.enclosing_block:
+            x = _mb.cast(x=feature_var, dtype=to_dtype_str, name=feature_desc.name + f"_to_{to_dtype_str}")
+            x.op.enclosing_block.replace_uses_of_var_after_op(
+                anchor_op=x.op,
+                old_var=feature_var,
+                new_var=x,
+                force_replace=True,
+            )
+            feature_desc.name = x.name
+            feature_desc.type.multiArrayType.dataType = to_type
+
+    if not isinstance(ml_model, _ct.models.MLModel):
+        raise ValueError(f"input model must be of type ct.models.MLModel, actual type is {type(ml_model)})")
+
+    model_spec = ml_model.get_spec()
+
+    model_type = model_spec.WhichOneof("Type")
+    if model_type != "mlProgram":
+        raise ValueError(f"input model must be an mlProgram, actual model type is {model_type}")
+
+    if not isinstance(from_type, int):
```

Comment:
For these two probably better to just `from_type`/`to_type` is `in (ArrayFeatureType.FLOAT16, ArrayFeatureType.FLOAT32)`